File: apps/web/src/sync/home-stats.test.ts
import type { FeedEntry, HomeSnapshot, HomeStats } from "@marshal/protocol";
import { unwrap } from "solid-js/store";
import { describe, expect, it } from "vitest";
import { golden } from "~/data/testing/golden";
import { contextOf, createTestMarshal } from "~/testing/test-store";
import { applyHomeStatsEvent, homeStatsSyncer } from "./home-stats";

// Section S19a: the stored numbers behind the Home charts (docs/backend-checklist.md B2.3). The
// syncer reads the daemon's range once and keeps the day a finished card changed current from the
// activity.created events of the home topic.

const stats = golden<HomeStats>("home-stats");
const created = golden<{ entry: FeedEntry }>("activity-created");

const snapshot: HomeSnapshot = {
  needs: [],
  awake: [],
  tiles: { needs: 0, working: 0, mergedToday: 0 },
  stats,
  serverTime: "2026-09-30T12:00:00.000Z",
};

/** The numbers one day's row is given. */
interface Counts {
  cardsFinished: number;
  merges?: number;
  ciFailures?: number;
}

/** The day an activity.created event carries, in the wire's own shape. */
function dayOf(counts: Counts) {
  return {
    day: "2026-09-30T00:00:00.000Z",
    cardsFinished: counts.cardsFinished,
    merges: counts.merges ?? 0,
    ciFailures: counts.ciFailures ?? 0,
    costMicros: 0,
  };
}

/** One activity.created event for a project's own day. */
function eventOf(projectId: string, counts: Counts) {
  return {
    seq: 1,
    topic: "home",
    type: "activity.created",
    at: "2026-09-30T12:00:00.000Z",
    data: { entry: { ...created.entry, projectId }, day: dayOf(counts) },
  } as never;
}

describe("the Home numbers section", () => {
  it("is section S19a, follows the home topic, and reads the activity events", () => {
    expect(homeStatsSyncer.section).toBe("S19a");
    expect(homeStatsSyncer.topics).toEqual(["home"]);
    expect(homeStatsSyncer.onEvent).toBeTypeOf("function");
  });

  it("maps the daemon's stored range into the store, one day per point", () => {
    const ctx = contextOf(createTestMarshal());
    homeStatsSyncer.apply(ctx, snapshot);
    expect(ctx.S.stats.range).toBe(7);
    expect(ctx.S.stats.days.map((day) => day.day)).toEqual([
      Date.UTC(2026, 8, 24),
      Date.UTC(2026, 8, 25),
      Date.UTC(2026, 8, 26),
      Date.UTC(2026, 8, 27),
      Date.UTC(2026, 8, 28),
      Date.UTC(2026, 8, 29),
      Date.UTC(2026, 8, 30),
    ]);
    expect(ctx.S.stats.days[6]).toMatchObject({ cardsFinished: 3, merges: 3 });
    expect(ctx.S.stats.projects.map((project) => project.projectId)).toEqual(["api", "web"]);
  });

  it("sets the whole day of the project a card finished in, and the totals with it", () => {
    const ctx = contextOf(createTestMarshal());
    homeStatsSyncer.apply(ctx, snapshot);
    // The event carries the project's own day as it is now, not a difference.
    applyHomeStatsEvent(ctx, eventOf("api", { cardsFinished: 4, merges: 4, ciFailures: 1 }));

    const api = ctx.S.stats.projects.find((project) => project.projectId === "api");
    expect(api?.days[6]).toEqual({
      day: Date.UTC(2026, 8, 30),
      cardsFinished: 4,
      merges: 4,
      ciFailures: 1,
      costMicros: 0,
    });
    // The totals are the projects added together: api's 4 and web's 2.
    expect(ctx.S.stats.days[6]).toEqual({
      day: Date.UTC(2026, 8, 30),
      cardsFinished: 6,
      merges: 6,
      ciFailures: 1,
      costMicros: 0,
    });
    // A day the event did not touch keeps what the snapshot gave it.
    expect(ctx.S.stats.days[0]).toMatchObject({ cardsFinished: 2, merges: 2 });
  });

  it("changes nothing when the same event is applied twice, or after a reload", () => {
    const ctx = contextOf(createTestMarshal());
    homeStatsSyncer.apply(ctx, snapshot);
    const event = eventOf("api", { cardsFinished: 4, merges: 4, ciFailures: 1 });
    applyHomeStatsEvent(ctx, event);
    const once = unwrap(ctx.S.stats);
    applyHomeStatsEvent(ctx, event);
    expect(ctx.S.stats).toEqual(once);
  });

  it("starts a project's series when the project has no stored day yet", () => {
    const ctx = contextOf(createTestMarshal());
    homeStatsSyncer.apply(ctx, snapshot);
    applyHomeStatsEvent(ctx, eventOf("billing", { cardsFinished: 1, merges: 1 }));

    const billing = ctx.S.stats.projects.find((project) => project.projectId === "billing");
    expect(billing?.days).toHaveLength(7);
    expect(billing?.days[0]).toEqual({
      day: Date.UTC(2026, 8, 24),
      cardsFinished: 0,
      merges: 0,
      ciFailures: 0,
      costMicros: 0,
    });
    expect(billing?.days[6]).toMatchObject({ cardsFinished: 1, merges: 1 });
    // The totals now count the new project too.
    expect(ctx.S.stats.days[6]).toMatchObject({ cardsFinished: 4 });
  });

  it("ignores a change that comes before the range was loaded, since the load brings it", () => {
    const ctx = contextOf(createTestMarshal());
    const before = ctx.S.stats;
    applyHomeStatsEvent(ctx, eventOf("api", { cardsFinished: 4, merges: 4 }));
    expect(ctx.S.stats).toBe(before);
  });

  it("ignores an entry that belongs to no project, and a row that changed no numbers", () => {
    const ctx = contextOf(createTestMarshal());
    homeStatsSyncer.apply(ctx, snapshot);
    const before = unwrap(ctx.S.stats);
    applyHomeStatsEvent(ctx, {
      seq: 2,
      topic: "home",
      type: "activity.created",
      at: "2026-09-30T12:00:00.000Z",
      data: { entry: { ...created.entry, projectId: null }, day: dayOf({ cardsFinished: 1 }) },
    } as never);
    applyHomeStatsEvent(ctx, {
      seq: 3,
      topic: "home",
      type: "activity.created",
      at: "2026-09-30T12:00:00.000Z",
      data: { entry: created.entry, day: null },
    } as never);
    applyHomeStatsEvent(ctx, { type: "card.moved", data: { card: {} } } as never);
    expect(ctx.S.stats).toEqual(before);
  });
});

File: daemon/internal/protocol/home_stats_test.go
package protocol_test

import (
	"testing"
	"time"

	"github.com/khanblair/marshal/daemon/internal/protocol"
	"github.com/khanblair/marshal/daemon/internal/testutil"
)

// The stored Home numbers and the activity stream (docs/backend-checklist.md B2.3, sections S19a
// and S20, inventory N17). A Go test writes the golden files and the TypeScript tests read them, so
// the two sides cannot drift apart.

// homeDay is midnight at the start of the day `daysAgo` days before the sample day, which is what a
// stored day is and what a chart's axis is built from.
func homeDay(daysAgo int) protocol.Timestamp {
	midnight := time.Date(2026, time.September, 30, 0, 0, 0, 0, time.UTC)
	return protocol.NewTimestamp(midnight.AddDate(0, 0, -daysAgo))
}

// homeStatDays is one week of days with the projects added together, oldest first. The sums of the
// two per-project series below are exactly these, so a client that adds the projects up gets the
// same chart as one that draws the totals.
func homeStatDays() []protocol.HomeStatDay {
	return []protocol.HomeStatDay{
		{Day: homeDay(6), CardsFinished: 2, Merges: 2},
		{Day: homeDay(5), CardsFinished: 3, Merges: 3},
		{Day: homeDay(4), CardsFinished: 1, Merges: 1, CIFailures: 1},
		{Day: homeDay(3), CardsFinished: 4, Merges: 3},
		{Day: homeDay(2), CardsFinished: 2, Merges: 2},
		{Day: homeDay(1), CardsFinished: 5, Merges: 4, CIFailures: 2},
		{Day: homeDay(0), CardsFinished: 3, Merges: 3},
	}
}

// homeStatProjects is the same week, one series per project, for the chart that draws a line per
// project.
func homeStatProjects() []protocol.HomeProjectStats {
	return []protocol.HomeProjectStats{
		{
			ProjectID: "api",
			Days: []protocol.HomeStatDay{
				{Day: homeDay(6), CardsFinished: 1, Merges: 1},
				{Day: homeDay(5), CardsFinished: 2, Merges: 2},
				{Day: homeDay(4), CIFailures: 1},
				{Day: homeDay(3), CardsFinished: 2, Merges: 2},
				{Day: homeDay(2), CardsFinished: 1, Merges: 1},
				{Day: homeDay(1), CardsFinished: 3, Merges: 2, CIFailures: 2},
				{Day: homeDay(0), CardsFinished: 1, Merges: 1},
			},
		},
		{
			ProjectID: "web",
			Days: []protocol.HomeStatDay{
				{Day: homeDay(6), CardsFinished: 1, Merges: 1},
				{Day: homeDay(5), CardsFinished: 1, Merges: 1},
				{Day: homeDay(4), CardsFinished: 1, Merges: 1},
				{Day: homeDay(3), CardsFinished: 2, Merges: 1},
				{Day: homeDay(2), CardsFinished: 1, Merges: 1},
				{Day: homeDay(1), CardsFinished: 2, Merges: 2},
				{Day: homeDay(0), CardsFinished: 2, Merges: 2},
			},
		},
	}
}

// sampleStats is the chart data of a seven-day Home answer.
func sampleStats() protocol.HomeStats {
	return protocol.HomeStats{
		Range:    7,
		From:     homeDay(6),
		To:       homeDay(0),
		Days:     homeStatDays(),
		Projects: homeStatProjects(),
	}
}

// The stored numbers behind the Home charts, over the range the route asked for.
func TestHomeStatsGolden(t *testing.T) {
	testutil.Golden(t, "home-stats", sampleStats())
}

// A range with nothing stored is a list of zeroed days, never null: a chart with no history draws a
// flat line rather than nothing at all.
func TestHomeStatsWithNothingStored(t *testing.T) {
	empty := protocol.HomeStats{
		Range:    30,
		From:     homeDay(29),
		To:       homeDay(0),
		Days:     []protocol.HomeStatDay{},
		Projects: []protocol.HomeProjectStats{},
	}
	testutil.Golden(t, "home-stats-empty", empty)
}

// One row of the activity stream and the page the view-all list reads. Every kind of subject is in
// the page: an entry about a card, one that belongs to no project, and one about a scheduled job.
func TestHomeActivityPageGolden(t *testing.T) {
	web := "web"
	api := "api"
	cardEntry := protocol.FeedEntry{
		ID:        "01M3C107JB041061050R3GG28A",
		Kind:      protocol.FeedKindMerge,
		Text:      "#110 Fix flaky login e2e test merged into main",
		ProjectID: &web,
		At:        sampleDate(),
		CardID:    "01M3C107JB041061050R3GG28B",
		CardKey:   "web#110",
	}
	briefEntry := protocol.FeedEntry{
		ID:    "01M3C107JB041061050R3GG28C",
		Kind:  protocol.FeedKindBrief,
		Text:  "Morning brief is ready",
		At:    protocol.NewTimestamp(time.Date(2026, time.September, 30, 9, 0, 0, 0, time.UTC)),
		JobID: "s1",
	}
	ciEntry := protocol.FeedEntry{
		ID:        "01M3C107JB041061050R3GG28D",
		Kind:      protocol.FeedKindCI,
		Text:      "CI passed on marshal/39-slog",
		ProjectID: &api,
		At:        protocol.NewTimestamp(time.Date(2026, time.September, 30, 8, 30, 0, 0, time.UTC)),
		CardID:    "01M3C107JB041061050R3GG28E",
		CardKey:   "api#39",
	}
	page := protocol.Page[protocol.FeedEntry]{
		Items:      []protocol.FeedEntry{cardEntry, briefEntry, ciEntry},
		NextCursor: "eyJzZXEiOjF9",
		ServerTime: sampleDate(),
	}
	testutil.Golden(t, "home-activity", page)
}

// The event that tells a client a row was appended, with the project's own day the row left behind
// it: the whole day, so applying the event twice changes nothing.
func TestActivityCreatedEventDataGolden(t *testing.T) {
	web := "web"
	entry := protocol.FeedEntry{
		ID:        "01M3C107JB041061050R3GG28A",
		Kind:      protocol.FeedKindMerge,
		Text:      "#110 Fix flaky login e2e test merged into main",
		ProjectID: &web,
		At:        sampleDate(),
		CardID:    "01M3C107JB041061050R3GG28B",
		CardKey:   "web#110",
	}
	today := protocol.HomeStatDay{Day: homeDay(0), CardsFinished: 3, Merges: 3}
	testutil.Golden(t, "activity-created", protocol.ActivityCreatedEventData{Entry: entry, Day: &today})
}

// An entry that does not change the numbers carries no day at all, so a client knows there is
// nothing to write into its chart.
func TestActivityCreatedEventDataWithoutADay(t *testing.T) {
	entry := protocol.FeedEntry{
		ID:    "01M3C107JB041061050R3GG28C",
		Kind:  protocol.FeedKindBrief,
		Text:  "Morning brief is ready",
		At:    sampleDate(),
		JobID: "s1",
	}
	testutil.Golden(t, "activity-created-no-day", protocol.ActivityCreatedEventData{Entry: entry, Day: nil})
}
